[feat]: [CDS-74808]: updating traffic routing docs with more information (#6002)
harness/developer-hub: docs/continuous-delivery/deploy-srv-diff-platforms/kubernetes/cd-k8s-ref/traffic-shifting-step.md

Specify your configuration type here. Currently there are two choices:

> * **New Config**: Select this option if you want to specify a new configuration for traffic routing in this step. If configuring this in BlueGreen deployment step or Canary step this option implicitly assumed. This option will create a new resource(s).

Reviewer: ```suggestion
* **New Config**: Select this option if you want to specify a new configuration for traffic routing in this step. If configuring this in a BlueGreen deployment step or a Canary step this option is implicitly assumed. This option will create a new resource(s).
```

* **Inherit**: Select this option if you want the traffic routing step to inherit a configuration from a previous Blue Green, Canary, or Traffic Routing step. This option will patch existing resources. 

### Provider

Specify your service mesh provider. Harness currently supports Service Mesh Interface (SMI) and Istio. 

> Each provider will have some common configuration options and some provider specific ones. We have listed all configuration options for each provider. Please take at the one relevant for you.

Reviewer: ```suggestion
Each provider will have some common configuration options and some provider specific ones. We have listed all configuration options for each provider. Please look at the one relevant for you.
```

              :::note
              If the total weights for all host destinations is not equal to 100, the weight values will be normalized into a percentage, and the pipeline will run with a warning.
              :::

#### Istio - New Config option

Before you begin, make sure you have an understanding of Istio and how it works by referring to [their website](https://istio.io/latest/about/service-mesh/).

* **Parameters**:
    * **Resource Name:** This name will be used to generate a kubernetes name for traffic resources. Hence the name needs to be kubernetes resource name compliant. 
    * **Hosts:** Specify one or more host names. This is specific to Istio, please take a look [here](https://istio.io/latest/docs/concepts/traffic-management/#the-hosts-field).
    * **Gateways:** Specify one or more gateway names. This is specific to Istio, please take a look [here](https://istio.io/latest/docs/reference/config/networking/gateway/).
    * **Routes**: Currently, Harness supports only the `http` route type.
        * **Route type**: Currently, Harness supports only the `http` route type - for http traffic.

>         * **Route name**: This will be used in generating kubernetes resources so the value should kubernetes name compliant. This should also be unique within the pipeline workflow as it is used as ID for updating purposes (described below)

Reviewer: ```suggestion
        * **Route name**: This will be used in generating kubernetes resources so the value should be kubernetes name compliant. This should also be unique within the pipeline workflow as it is used as an ID for updating purposes (described below).
```

        * **Route Rules**: This is the way incoming requests are filtered for the configured route destinations.
            * **uri**:

>                 * **Value**: Enter the value that you want matched against incoming request URI.

Reviewer: ```suggestion
                * **Value**: Enter the value that you want matched against the incoming request URI.
```

>                 * **Match Type**: Used to determine how URI of the incoming request is matched against the `value` config parameter. Available to choose one of three values (`exact`, `prefix`, `regex`).

Reviewer: ```suggestion
                * **Match Type**: Used to determine how the URI of the incoming request is matched against the `value` config parameter. Choose from one of the three available values (`exact`, `prefix`, `regex`).
```

            * **method**:
                * **Value**: Used for matching the HTTP method of the incoming request.

>                 * **Match Type**: Used to determine how HTTP method of the incoming request is matched against the `value` config parameter. Available to choose one of three values (`exact`, `prefix`, `regex`).

Reviewer: ```suggestion
                * **Match Type**: Used to determine how the HTTP method of the incoming request is matched against the `value` config parameter. Choose from one of the three available values (`exact`, `prefix`, `regex`).
```

            * **headers**:
                * **Values**:
                    - **Key**: Specify the name of the request header
                    - **Value**: Specify the value of the header

>                     - **Match Type**: Specify which kind of matching should be done against the incoming header. Available to choose one of three values (`exact`, `prefix`, `regex`).

Reviewer: ```suggestion
                    - **Match Type**: Specify which kind of matching should be done against the incoming header. Choose from one of the three available values  (`exact`, `prefix`, `regex`).
```

    
            * **scheme**:
                * **Value**: Specify which scheme you want to match with.

>                 * **Match Type**: Specify which kind of matching should be done against the incoming request scheme. Available to choose one of three values (`exact`, `prefix`, `regex`).

Reviewer: ```suggestion
                * **Match Type**: Specify which kind of matching should be done against the incoming request scheme. Choose from one of the three available values (`exact`, `prefix`, `regex`).
```

            * **authority**:
                * **Value**: Specify which authority you want to match with.

>                 * **Match Type**: Specify which kind of matching should be done against the incoming request authority. Available to choose one of three values (`exact`, `prefix`, `regex`).

Reviewer: ```suggestion
                * **Match Type**: Specify which kind of matching should be done against the incoming request authority. Choose from one of the three available values (`exact`, `prefix`, `regex`).
```

            * **port**: 

>                 * **Value**: Specify which port you want to match incoming request port with.

Reviewer: ```suggestion
                * **Value**: Specify which port you want to match the incoming request port with.
```

>         * **Destinations**: This is locations to which the filtered request will be routed to

Reviewer: ```suggestion
        * **Destinations**: The filtered requests will be routed to these locations.
```

            * **Host**: Should be the name of the Kubernetes service resource.

>               :::note 

Reviewer: What are you trying to improve in this note or what information was missing originally? I'm trying to understand what different information you want to convey.

Author: I just changed the indentation of it. Pushed it under the `host` and `weight`.

Reviewer: I mean it used to be this:

  :::note 
    For Blue/Green deployments we recommend using the placeholder `stable` and `stage` resource names.

    For Canary deployments we recommend using the placeholder `stable` and `canary` resource names.

    These names will be replaced by the actual Kubernetes service resource names which were generated during the B/G or canary deployment.
    :::
    
    and now its this:
    
    :::note
    *Only applicable when using the Blue/Green deployment step* - You can use placeholder `stable` and `stage` resource names. Which would effectively be replaced with the name of the stable and stage services, respectively. 

              *Only applicable when using the Canary deployment step* You can use the placeholder `stable` and `canary` resource names. Which would effectively be replaced with the name of the stable and canary services, respectively.
              :::

Reviewer: :::note
For Blue/Green deployment steps we recommend using the placeholder `stable` and `stage` resource names.

For Canary deployment steps we recommend using the placeholder `stable` and `canary` resource names.

These names will be replaced by the actual Kubernetes service resource names which were generated during the B/G or canary deployment.
:::

              :::note
              If the total weights for all host destinations is not equal to 100, the weight values will be normalized into a percentage, and the pipeline will run with a warning.
              :::

#### SMI and Istio - Inherit option
This options provides a way to update an existing traffic routing configuration destination's weights. With this step configuration there are two logical parts.
First one is to configure the `Route Name` and second one is to configure destination(s) which you want to update. Route name is an ID for us to know which traffic routing resource we need to update. The value of the route name should match one of the route names defined during the `New config` step configuration.
You can configure one or more destinations and their weights. 
You should also be aware that in case you configure lower number of destinations then the route contains, the update of the weights will occur, but it will respect [0-100] boundaries.

Example:
In original setup you have configured 3 destinations for a route:
```
Destination1 -> Host: svc1, Weight: 60
Destination2 -> Host: svc2, Weight: 30
Destination3 -> Host: svc3, Weight: 10
```

> Now in the update step you configured to update only one destination for that same route. Let's say you want to update Destination `svc1` and what to set its weight to `40`

Reviewer: ```suggestion
Now in the update step you configured to update only one destination for that same route. Let's say you want to update Destination `svc1` and you want to set its weight to `40`
```

> This will lead to having the remaining 60 (max limit of 100 - configured update weight of 40) to be split amongst the remaining destination, but keeping the same ratio they have amongst themselves.

Reviewer: ```suggestion
This will lead to having the remaining 60 (max limit of 100 - configured update weight of 40) to be split amongst the remaining destinations, but keeping the same ratio they have amongst themselves.
```
